tests/test_picking_batch.py

```diff
@@ -305,3 +307,215 @@ def test14_is_valid_location_dest_failure_unknown_location(self):
+
+        self.assertEqual(picking.state, 'cancel')
+        self.assertEqual(batch.state, 'done')
+        self.assertEqual(internal_picking.picking_type_id, self.picking_type_internal)
+        self.assertEqual(internal_picking.state, 'assigned')
+        self.assertEqual(batch.state, 'done')
+
+    def test16_unpickable_item_single_move_line_has_package_success_picking_type_specified(self):  # noqa
+        """
+        Tests that the picking is cancelled and the specified picking type
+        is created for the unpickable stock
+        """
+        picking, batch = self._create_valid_batch_for_location_tests()
+        move_line_id = picking.move_line_ids[0].id
+        reason = 'missing item'
+
+        # It doesn't actually matter what the picking type is for this test.
+        # The important thing is that the new picking is of type picking_type
+        picking_type = self.picking_type_pick
+        batch.unpickable_item(move_line_id=move_line_id,
+                              reason=reason,
+                              picking_type_id=picking_type.id)
+        internal_picking = self.package_one.find_move_lines().picking_id  # noqa
+
+        self.assertEqual(picking.state, 'cancel')
+        self.assertEqual(batch.state, 'done')
+        self.assertEqual(internal_picking.picking_type_id, picking_type)
+        self.assertEqual(internal_picking.state, 'assigned')
+
+    def test17_unpickable_item_move_line_not_found_done(self):
+        """
+        Tests that a ValidationError is raised if the move_line_id cannot be
+        found
+        """
+        picking, batch = self._create_valid_batch_for_location_tests()
+        picking.update_picking(force_validate=True,
+                               location_dest_id=self.test_output_location_01.id)  # noqa
+        reason = 'missing item'
+
+        with self.assertRaises(ValidationError) as err:
+            batch.unpickable_item(move_line_id=999,
+                                  reason=reason,
+                                  picking_type_id=None)
+            self.assertEqual(err.exception.name,
+                             'Cannot find the operation')
+
+    def test18_unpickable_item_wrong_batch(self):
+        """
+        Tests that a ValidationError is raised if the move_line_id is not on
+        the Batch that we requested.
+        """
+        picking, batch = self._create_valid_batch_for_location_tests()
+        # Create a quant and picking for a different package
+        self.create_quant(self.apple.id, self.test_location_01.id, 4,
+                          package_id=self.package_two.id)
+        different_picking = self.create_picking(
+            self.picking_type_pick,
+            products_info=self.pack_4apples_info,
+            confirm=True,
+            assign=True)
+        move_line_id = different_picking.move_line_ids[0].id
+        reason = 'missing item'
+        with self.assertRaises(ValidationError) as err:
```

Comment:
Nitpick: an empty line above the `with` block makes it easier to distinguish the preparation code from the actual test condition.